SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorThreadChecker.java

```diff
@@ -101,13 +111,51 @@ private void checkThreads() {
         continue;
       }
 
-      if (usedThreads > maxThreads) {
+      if (usedThreads.get() > maxThreads) {
         taskProcess.getTask().getLog().info("{} using too many threads: {} (max {})", taskProcess.getTask().getTaskId(), usedThreads, maxThreads);
 
-        taskProcess.getTask().markKilledDueToThreads(usedThreads);
-        KillState killState = monitor.requestKill(taskProcess.getTask().getTaskId());
+        if (configuration.getRunShellCommandBeforeKillDueToThreads().isPresent()) {
+          SingularityTaskShellCommandRequest shellRequest = new SingularityTaskShellCommandRequest(
+              SingularityTaskId.valueOf(taskProcess.getTask().getTaskId()),
+              Optional.absent(),
+              System.currentTimeMillis(),
+              configuration.getRunShellCommandBeforeKillDueToThreads().get()
+          );
+
+          SingularityExecutorShellCommandUpdater updater = new SingularityExecutorShellCommandUpdater(
+              objectMapper, shellRequest, taskProcess.getTask()
+          );
+
+          SingularityExecutorShellCommandRunner shellRunner = new SingularityExecutorShellCommandRunner(shellRequest, configuration, taskProcess.getTask(),
+              taskProcess, monitor.getShellCommandExecutorServiceForTask(taskProcess.getTask().getTaskId()), updater);
+
+          Futures.addCallback(shellRunner.start(), new FutureCallback<Integer>() {
```

Comment:
curious, any reason for Futures vs CompletableFuture here?

Reply:
Mostly just staying consistent with shell commands kicked off elsewhere in the codebase, which all use the same `Futures.addCallback()` style here